skrub/_table_vectorizer.py

```diff
@@ -555,16 +555,14 @@ def _apply_cast(self, X: pd.DataFrame) -> pd.DataFrame:
         for col, dtype in self.types_.items():
             # if categorical, add the new categories to prevent
             # them to be encoded as nan
-            if pd.api.types.is_categorical_dtype(dtype):
+            if isinstance(dtype, pd.CategoricalDtype):
                 known_categories = dtype.categories
                 new_categories = pd.unique(X[col])
                 dtype = pd.CategoricalDtype(
                     categories=known_categories.union(new_categories)
                 )
                 self.types_[col] = dtype
-            X[col] = X[col].astype(
-                dtype
-            )  # we dont use .loc (https://github.com/pandas-dev/pandas/issues/53729)
+        X = X.astype(self.types_)
```

Comment:
This one has been modified at the same time and might not be raising a future warning anymore but I still think this solution is more efficient because we convert the dtypes all at once.